### ISSUE TITLE
[Proposal] Create matcher to match request params and headers

### ISSUE THREAD

**Issue by leomao10** [bblimke/webmock#892]
The current matcher `have_been_requested` is quite awesome, however, I found that the code I endup with would be something like this:

```ruby
    let(:email) { FFaker::Internet.email }
    let(:org_id) { SecureRandom.hex(24) }
    let(:url) { "#{ENV['THIRD_PARTY_HOST']}/authentication/resend_invite" }
    let!(:resend_invite_request) do
      stub_request(:post, url)
        .with(
          body: {
            email: email,
            orgId: org_id
          },
          headers: {"Authorization" => "Bearer #{registration_token}"}
        ).to_return(body: {}.to_json, status: 200)
    end
    
    before do
      method_make_api_call
    end

    it { expect(resend_invite_request).to have_been_requested }
```
But it is not really clear that what I want to test.

What I really care is, I would expect `method_make_api_call` would trigger api with certain params, so the ideal syntax would be:

```ruby
    let(:email) { FFaker::Internet.email }
    let(:org_id) { SecureRandom.hex(24) }
    let(:url) { "#{ENV['THIRD_PARTY_HOST']}/authentication/resend_invite" }

    let!(:resend_invite_request) do
      stub_request(:post, url).to_return(body: {}.to_json, status: 200)
    end
    
    before do
      method_make_api_call
    end

    it "call resend_invite request with email and org_id" do 
      expect(resend_invite_request).to have_been_requested.with(
          body: {
            email: email,
            orgId: org_id
          },
          headers: {"Authorization" => "Bearer #{registration_token}"}
        )
    end
```

**Comment by bblimke** [comment]
@leomao10 thank you for submitting this issue. 

Why not to add the params to the request stub?

I like how your proposed code is readable. My concern would be not to confuse people about what is the 'right' way to use that matcher.

**Comment by jasonkarns** [comment]
I agree with the OP and found this proposal when being confused why `have_been_requested` doesn't respond to `with`.

The AAA-style is the goal here. Arrange, sets up the minimal bits necessary for the stub. If testing an http adapter for instance, or an end-to-end test, I would likely have a single stub that is configured to match against a specific domain and http-method combo. Most likely, this stub would be a let-var that is the same for all specs.

The Act, of course, would be hitting the SUT that triggers all the calls.

The Assert, then would be matching against the actual things sent to the endpoint.


Keeping these separate has a number of benefits. For one, it makes the failure message more useful. The failure message can ignore bits about the http method or url, and instead print a diff output of the body received vs expected.

Secondly, it keeps the parts that are _unique_ about the test colocated with the test. Most commonly, the only thing that varies between different aspects of the SUT (under different contexts or different inputs) are that the params or body sent over the wire differ. If these must be declared outside of the test and up above, then the things that make that test unique are no longer close to the test itself. Ideally, we would be highlighting the things _within_ the test that make it unique (why it exists in the first place), while all the things that are irrelevant to the test would be configured elsewhere (in before hooks, or let vars higher up).

This would also make the http expectations from webmock consistent with how rspec-mocks itself sets up mocks. Arrange the stub via allow (likely up in a before or whatever), and then post-action assertions are then able to specify argument matchers and whatnot within the test itself.

**Comment by jasonkarns** [comment]
FWIW, a pattern I've been sliding into is:

```
  let(:some_api) { stub_request(:post, /host/) } # this stubs the endpoint for the entire test file since some of the tests may hit it or may not. but the _expectations_ are test-specific

  # some describes or contexts

  # maybe these tests here `expect(some_api).to have_been_requested` or maybe not at all

  # then a test that cares precisely about _how_ the api was hit
    specify do
      some_api.with(body: {
        param1: some_matcher
      })

      subject.do_the_thing

      expect(some_api).to have_been_requested
    end
  end
```

This sets up the "allow" stubbing for the api at the top level; keeps the test-specific expectations colocated within the test, and also avoids any rubocop complaints about `let!(:some_api)` ;)
